Parser: Support multiple attribute type, reject invalid

aduth opened the pull request:
Blocker for: #9634

This pull request seeks to enhance the block parser to support multiple `type`, in accordance with JSON schema. This is planned to be used in #9634 to support a nullable string for `align` attribute, representable in JSON schema as multiple type `[ "string", "null" ]`.

In doing so, it needs to account for how to handle existing behavior of attribute value casting, necessary for attributes sourced ambiguously as strings (e.g. as HTML attribute values). With these changes, casting only occurs for ambiguous string sources of a singular attribute type. For any other attribute, we now reject the attribute value (adopting behavior as if `undefined`, accounting for default value). This may be considered a breaking change, but only in cases where values were assigned of the wrong type, relying on casting to correct. More often, this should help capture unintended serialization (e.g. objects serialized as casted string `[object Object]`).

**Testing instructions:**

Verify unit tests pass.

```
npm test
```

aduth committed:
Parser: Support multiple attribute type, reject invalid

aduth committed:
Parser: Respect default of undefined ambiguous string source attribute

aduth committed:
Blocks: Extract ID as number from raw image

youknowriad reviewed (approved):
LGTM 👍

jorgefilipecosta commented on packages/blocks/src/api/parser.js:
```diff
@@ -49,6 +61,89 @@ export const toBooleanAttributeMatcher = ( matcher ) => flow( [
+
+	return (
+		isStringSource( source ) &&
+		typeof type === 'string'
```
Should we check if type === 'string' and not typeof type === 'string'?

aduth replied:
No, because this is how we differentiate between single and multiple types. I can extract a variable to make this clearer.

aduth committed:
Testing: Fix gallery columns fixtures source

aduth committed:
Parser: Refactor type tests to use typeof

aduth committed:
Parser: Clarify ambiguous string source by explaining variables

aduth committed:
Parser: Deprecate type coercion

It's unclear for which string sources should be considered ambiguous and allow casting, in addition to being inconsistently enforced. Rather than defaulting to leniency, we should enforce the type broadly, reintroducing leniency later if noted issues and need for multiple type support can be reconciled.
